enclave-runtime/src/utils.rs

```diff
@@ -38,6 +37,45 @@ pub fn write_slice_and_whitespace_pad(writable: &mut [u8], data: Vec<u8>) {
 	right.iter_mut().for_each(|x| *x = 0x20);
 }
 
+/// Helper trait to transform the sgx-ffi pointers to any type that implements
+/// `parity-scale-codec::Decode`
+pub unsafe trait DecodeRaw {
+	/// the type to decode into
+	type Decoded: Decode;
+
+	unsafe fn decode_raw<'a, T>(data: *const T, len: usize) -> Result<Self::Decoded, codec::Error>
+	where
+		T: 'a,
+		&'a [T]: Input;
+}
+
+unsafe impl<D: Decode> DecodeRaw for D {
+	type Decoded = D;
+
+	unsafe fn decode_raw<'a, T>(data: *const T, len: usize) -> Result<Self::Decoded, Error>
+	where
+		T: 'a,
+		&'a [T]: Input,
```

Comment:
Seems like a really useful piece of code here 👍 

So for my own understanding:
What is the `Input` trait?

Reply:
It is the trait bound that the parity_scale_codec needs. Here would be the definition: https://github.com/paritytech/parity-scale-codec/blob/166d748abc1e48d74c528e2456fefe6f3c48f256/src/codec.rs#L64

Reply:
ah I see, thanks!